File: Progressions/progressions.js
class ArithmeticProgression{
    
    #d = 0;
    #a = [];
    
    constructor(start = 0, d = 0){
        this.#a.push(start);
        this.#d = d;
    }

    get sequence(){
        return this.#a;
    }

    get difference(){
        return this.#d;
    }

    set startElement(start){
        this.#a = [start];
    }

    set difference(d){
        this.#a = [this.#a[0]];
        this.#d = d;
    }

    getNth(n){
        return this.#a[0] + (n - 1) * this.#d;
    }

    countToNth(n){
        for(let _ = this.#a.length; _ < n; _ ++){
            this.#a.push(this.#a[this.#a.length - 1] + this.#d);
        }
    }

    getToNth(n){
        this.countToNth(n);
        return this.#a;
    }

    getSumToNth(n){
        return ~~(n * (this.#a[0] + this.getNth(n)) / 2);
    }
}

class GeometricProgression{
    
    #r = 1;
    #b = [];
    
    constructor(start = 1, r = 1){
        if(start == 0 || r == 0){
            console.log("startElement or ratio can't be 0");
            return; 
        }
        this.#b.push(start);
        this.#r = r;
    }

    get sequence(){
        return this.#b;
    }

    get ratio(){
        return this.#r;
    }

    set startElement(start){
        if(start == 0){
            console.log("startElement can't be 0");
            return; 
        }
        this.#b = [start];
    }

    set ratio(r){
        if(r == 0){
            console.log("ratio can't be 0");
            return; 
        }
        this.#b = [this.#b[0]];
        this.#r = r;
    }

    getNth(n){
        return this.#b[0] * (this.#r ** (n - 1));
    }

    countToNth(n){
        for(let _ = this.#b.length; _ < n; _ ++){
            this.#b.push(this.#b[this.#b.length - 1] * this.#r);
        }
    }

    getToNth(n){
        this.countToNth(n);
        return this.#b;
    }

    getSumToNth(n){
        if(this.#r == 1) return n * this.#b[0];
        return ~~((this.#b[0] * ((this.#r ** n) - 1)) / (this.#r - 1));
    }

    getProdToNth(n){
        return ~~((this.#b[0] * this.get_nth(n)) ** (n / 2));
    }
}

class HarmonicProgression{
    
    #d = 0;
    #c = [];
    
    constructor(start = 1, d = 0){
        this.#c.push(start);
        this.#d = d;
    }

    get sequence(){
        return this.#c;
    }

    get difference(){
        return this.#d;
    }

    set startElement(start){
        this.#c = [start];
    }

    set difference(d){
        this.#c = [this.#c[0]];
        this.#d = d;
    }

    getNth(n){
        return ((this.#c[0] ** -1) + (n - 1) * this.#d) ** -1;
    }

    countToNth(n){
        for(let _ = this.#c.length; _ < n; _ ++){
            this.#c.push(((this.#c[this.#c.length - 1] ** -1) + this.#d) ** -1);
        }
    }

    getToNth(n){
        this.countToNth(n);
        return this.#c;
    }
}

File: Progressions/progressions.py
class ArithmeticProgression:

    __d = 0
    __a = []

    def __init__(self, start = 0, d = 0):
        self.__a.append(start)
        self.__d = d

    def get_sequence(self):
        return self.__a

    def get_difference(self):
        return self.__d

    def set_start_element(self, start):
        self.__a = [start]

    def set_difference(self, d):
        self.__a = [self.__a[0]]
        self.__d = d

    def get_nth(self, n):
        return self.__a[0] + (n - 1) * self.__d
    
    def count_to_nth(self, n):
        for _ in range(len(self.__a), n):
            self.__a.append(self.__a[-1] + self.__d)

    def get_to_nth(self, n):
        self.count_to_nth(n)
        return self.__a

    def get_sum_to_nth(self, n):
        return n * (self.__a[0] + self.get_nth(n)) // 2

class GeometricProgression:

    __r = 1
    __b = []

    def __init__(self, start = 1, r = 1):
        if start == 0 or r == 0:
            print("start_element or ratio can't be 0")
            return 
        self.__b.append(start)
        self.__r = r

    def get_sequence(self):
        return self.__b

    def get_ratio(self):
        return self.__r

    def set_start_element(self, start):
        if start == 0:
            print("start_element can't be 0")
            return
        self.__b = [start]

    def set_ratio(self, r):
        if r == 0:
            print("ratio can't be 0")
            return
        self.__b = [self.__r[0]]
        self.__r = r

    def get_nth(self, n):
        return self.__b[0] * (self.__r ** (n - 1))
    
    def count_to_nth(self, n):
        for _ in range(len(self.__b), n):
            self.__b.append(self.__b[-1] * self.__r)

    def get_to_nth(self, n):
        self.count_to_nth(n)
        return self.__b

    def get_sum_to_nth(self, n):
        if self.__r == 1: return n * self.__b[0]
        return (self.__b[0] * ((self.__r ** n) - 1)) // (self.__r - 1)

    def get_prod_to_nth(self, n):
        return int((self.__b[0] * self.get_nth(n)) ** (n / 2))

class HarmonicProgression:

    __d = 0
    __c = []

    def __init__(self, start = 1, d = 0):
        self.__c.append(start)
        self.__d = d
    
    def get_sequence(self):
        return self.__с

    def get_difference(self):
        return self.__d

    def set_start_element(self, start):
        self.__с = [start]

    def set_difference(self, d):
        self.__с = [self.__с[0]]
        self.__d = d

    def get_nth(self, n):
        return ((self.__c[0] ** -1) + (n - 1) * self.__d) ** -1

    def count_to_nth(self, n):
        for _ in range(len(self.__c), n):
            self.__c.append(((self.__c[-1] ** -1) + self.__d) ** -1)

    def get_to_nth(self, n):
        self.count_to_nth(n)
        return self.__c